File: TraningAPI/StoryBoard/StoryBoard.m
//
//  StoryBoard.m
//  TraningAPI
//
//  Created by 高橋慧 on 2021/06/21.
//  Copyright © 2021 k-takahashi. All rights reserved.
//

#import <Foundation/Foundation.h>
#import "StoryBoard.h"
#import "PrefecturesListViewController.h"
#import "DetailsViewController.h"
#import "MenuViewController.h"
#import "LicenceListViewController.h"
#import "LicenceContentViewController.h"
@implementation StoryBoard
+ (PrefecturesListViewController *)performPrefecturesListView: (UIViewController *)from {
    Board view = prefecturesList;
    NSString *identifer = Identifer(view);
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
    PrefecturesListViewController *nextVC = [storyboard instantiateViewControllerWithIdentifier: identifer];
    return nextVC;
}
+ (DetailsViewController *)performDetailsView: (UIViewController *)from prefecturesID:(int)idNumber {
    Board view = details;
    NSString *identifer = Identifer(view);
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
    DetailsViewController *nextVC = [storyboard instantiateViewControllerWithIdentifier: identifer];
    nextVC.idNumber = idNumber;
    return nextVC;
}
+ (MenuViewController *)performMenuView: (UIViewController *)from {
    Board view = menuView;
    NSString *identifer = Identifer(view);
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
    MenuViewController *nextVC = [storyboard instantiateViewControllerWithIdentifier: identifer];
    return nextVC;
}
+ (LicenceListViewController *)performLicenceListView: (UIViewController *)from {
    LicenceListViewController *nextVC = [[LicenceListViewController alloc] init];
    return nextVC;
}
+ (LicenceContentViewController *)performLicenceContentView: (UIViewController *)form licencesID:(int)idNumber {
    LicenceContentViewController *nextVC = [[LicenceContentViewController alloc] init];
    nextVC.titles = Common.libraryTitle[idNumber];
    nextVC.content = Common.libraryContent[idNumber];
    return nextVC;
}
+ (void)push:(UIViewController *)from toViewController:(UIViewController *)to {
    [from.navigationController pushViewController:to animated:true ];
}
@end

File: TraningAPI/StoryBoard/StoryBoard.swift
//
//  StoryBoard.swift
//  TraningAPI
//
//  Created by 高橋慧 on 2021/04/16.
//  Copyright © 2021 k-takahashi. All rights reserved.
//

import UIKit
enum Board {

    case prefecturesList
    case details
    case menuView
    case licenceList
    case licenceContent
    
    var identifer: String{
        switch self {
        case .prefecturesList: return "PrefecturesListViewController"
        case .details:         return "DetailsViewController"
        case .menuView:        return "MenuViewController"
        case .licenceList:     return "LicenceListViewController"
        case .licenceContent:  return "LicenceContentViewController"
        }
    }
    
    func boardInit(id: Int) -> UIViewController?{
        switch self {
        case .prefecturesList:
            let nextVC = UIStoryboard(name: "Main", bundle: nil).instantiateViewController(withIdentifier: self.identifer) as? PrefecturesListViewController
            return nextVC
        case .details:
            let nextVC = UIStoryboard(name: "Main", bundle: nil).instantiateViewController(withIdentifier: self.identifer) as? DetailsViewController
            nextVC?.id = id
            return nextVC
        case .menuView:
            let nextVC = MenuViewController()
            return nextVC
        case .licenceList:
            let nextVC = LicenceListViewController()
            return nextVC
        case .licenceContent:
            let nextVC = LicenceContentViewController()
            nextVC.title = ConstEntity.LIBRARYTITLE[id]
            nextVC.content = ConstEntity.LIBRARYCONTENT[id]
            return nextVC
        }
    }
}

final class StoryBoard{
    //    MARK: - Screen Transition　画面遷移
    /// - Parameters:
    ///   - id :Prefecturesテーブルのid
    ///   - to:遷移先のViewController
    ///   - from:遷移元のViewController
    ///Show遷移
    static func perform(id:Int, to:Board, from:UIViewController){
        if let nextVC = to.boardInit(id: id){
            from.navigationController?.pushViewController(nextVC, animated: true)
        }
    }
    deinit {
          print("storyboard", #function)
      }
}
